File: frontend/src/api/checklist.ts
import { api } from './client'

export interface ChecklistStatus {
  checklist_done: boolean
  cleanliness_done_today: boolean
}

export interface VesselOption {
  value: string
  label: string
}

export interface VesselOptions {
  has_vessel: boolean
  options: VesselOption[]
}

export interface ComplianceRow {
  operator_name: string
  pump_station: string
  shift: string
  poured: number
  checklist_at: string | null
  start_audit_at: string | null
  transfer_audit_at: string | null
  end_audit_at: string | null
  start_expected: boolean
  transfer_expected: boolean
  end_expected: boolean
  /** A quick job on a pump other than the one the shift began on - no photos owed. */
  brief: boolean
  complete: boolean
}

export interface ComplianceOut {
  date: string
  rows: ComplianceRow[]
  /** "self" for an operator, "everyone" for a manager or admin. */
  scope: string
}

export const checklistApi = {
  compliance: (onDate: string, asOperator?: string) => {
    const params = new URLSearchParams({ on_date: onDate })
    if (asOperator) params.set('as_operator', asOperator)
    return api.get<ComplianceOut>(`/checklist/compliance?${params}`)
  },
  status: (station: string, shift: string) =>
    api.get<ChecklistStatus>(`/checklist/status?${new URLSearchParams({ station, shift })}`),
  vesselOptions: (station: string) =>
    api.get<VesselOptions>(`/checklist/vessel-options?${new URLSearchParams({ station })}`),
  submitCleanliness: (formData: FormData) =>
    api.postForm<{ ok: boolean }>('/checklist/cleanliness', formData),
  submit: (body: {
    station: string
    shift: string
    qr_checked: boolean
    materials_checked: boolean
    vessel_reactor_name?: string | null
  }) => api.post<{ ok: boolean }>('/checklist/submit', body),
  markAlreadyDone: (body: { station: string; shift: string; already_who: string }) =>
    api.post<{ ok: boolean }>('/checklist/mark-already-done', body),
}

File: api/schemas/checklist.py
from pydantic import BaseModel


class ChecklistStatus(BaseModel):
    checklist_done: bool
    cleanliness_done_today: bool


class VesselOption(BaseModel):
    value: str  # reactor_name - sent back on /submit as vessel_reactor_name
    label: str  # shared._vessel_label(v)


class VesselOptionsOut(BaseModel):
    has_vessel: bool
    options: list[VesselOption]


class ChecklistSubmitRequest(BaseModel):
    station: str
    shift: str
    qr_checked: bool
    materials_checked: bool
    vessel_reactor_name: str | None = None


class MarkAlreadyDoneRequest(BaseModel):
    station: str
    shift: str
    already_who: str


class ComplianceRow(BaseModel):
    """One operator at one pump on one shift, and which of their checks are
    on record. Times are ISO with a UTC offset, or null for "not done"."""
    operator_name: str
    pump_station: str
    shift: str
    poured: int
    checklist_at: str | None
    start_audit_at: str | None
    transfer_audit_at: str | None
    end_audit_at: str | None
    start_expected: bool
    transfer_expected: bool
    end_expected: bool
    # A quick job on a pump other than the one the shift began on (under
    # crud.BRIEF_VISIT_UNITS units): no photos are expected for it.
    brief: bool = False
    complete: bool


class ComplianceOut(BaseModel):
    date: str
    rows: list[ComplianceRow]
    # Only a manager sees other people's rows; an operator sees their own.
    scope: str
